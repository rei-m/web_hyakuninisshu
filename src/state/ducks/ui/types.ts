import { Action } from 'redux';
import { Color, Kimariji } from '@src/types';
import * as constants from './constants';

export type OPEN_KARUTAS_FILTER_TYPE = typeof constants.OPEN_KARUTAS_FILTER_NAME;

export type CLOSE_KARUTAS_FILTER_TYPE = typeof constants.CLOSE_KARUTAS_FILTER_NAME;

export type TOGGLE_KARUTAS_COLOR_TYPE = typeof constants.TOGGLE_KARUTAS_COLOR_NAME;

export type TOGGLE_KARUTAS_KIMARIJI_TYPE = typeof constants.TOGGLE_KARUTAS_KIMARIJI_NAME;

export interface OpenKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface CloseKarutasFilterAction extends Action {
  readonly type: CLOSE_KARUTAS_FILTER_TYPE;
}

export interface OKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface ToggleKarutasColorAction extends Action {
  readonly type: TOGGLE_KARUTAS_COLOR_TYPE;
  readonly payload: {
    readonly color: Color;
    readonly checked: boolean;
  };
}

export interface ToggleKarutasKimarijiAction extends Action {
  readonly type: TOGGLE_KARUTAS_KIMARIJI_TYPE;
  readonly payload: {
    readonly kimariji: Kimariji;
    readonly checked: boolean;
  };
}

export type Actions =
  | OpenKarutasFilterAction
  | CloseKarutasFilterAction
  | ToggleKarutasColorAction
  | ToggleKarutasKimarijiAction;

export interface KarutasFilter {
  readonly colors: Array<{ color: Color; checked: boolean }>;
  readonly kimarijis: Array<{ kimariji: Kimariji; checked: boolean }>;
}

export interface UiState {
  readonly karutasFilter: KarutasFilter & { readonly open: boolean };
}
