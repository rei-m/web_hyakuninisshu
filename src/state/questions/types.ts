import { Action } from 'redux';
import { Kimariji, Color, QuestionId, KarutaNo, Karuta } from '@src/domain/models';
import * as constants from './constants';
import { Meta, Payload } from '..';

type Condition<T> = {
  values: Array<T>;
};

export type RangeFromCondition = 1 | 11 | 21 | 31 | 41 | 51 | 61 | 71 | 81 | 91;
export const RangeFromCondition: Condition<RangeFromCondition> = {
  values: [1, 11, 21, 31, 41, 51, 61, 71, 81, 91],
};

export type RangeToCondition = 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 100;
export const RangeToCondition: Condition<RangeToCondition> = {
  values: [10, 20, 30, 40, 50, 60, 70, 80, 90, 100],
};

export type KimarijiCondition = null | Kimariji;
export const KimarijiCondition: Condition<KimarijiCondition> = {
  values: [null, 1, 2, 3, 4, 5, 6],
};

export type ColorCondition = null | Color;
export const ColorCondition: Condition<ColorCondition> = {
  values: [null, 'blue', 'pink', 'yellow', 'green', 'orange'],
};

export type KarutaStyleCondition = 'kanji' | 'kana';
export const KarutaStyleCondition: Condition<KarutaStyleCondition> = {
  values: ['kanji', 'kana'],
};

export type QuestionAnimCondition = 'none' | 'slow' | 'normal' | 'fast';
export const QuestionAnimCondition: Condition<QuestionAnimCondition> & {
  toDulation: (value: QuestionAnimCondition) => number;
} = {
  values: ['none', 'slow', 'normal', 'fast'],
  toDulation: (value: QuestionAnimCondition) => {
    switch (value) {
      case 'normal':
        return 0.6;
      case 'fast':
        return 0.3;
      case 'slow':
        return 1.0;
      default:
        return 0;
    }
  },
};

export type YomiFuda = Readonly<{
  karutaNo: KarutaNo;
  shoku: string;
  niku: string;
  sanku: string;
}>;

export type ToriFuda = Readonly<{
  karutaNo: KarutaNo;
  shiku: string;
  kekku: string;
}>;

export type StartTrainingAction = Action<typeof constants.START_TRAINING_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
    totalCount: number;
  }> &
  Meta<{
    rangeFrom: RangeFromCondition;
    rangeTo: RangeToCondition;
    kimariji: KimarijiCondition;
    color: ColorCondition;
    kamiNoKuStyle: KarutaStyleCondition;
    shimoNoKuStyle: KarutaStyleCondition;
    questionAnim: QuestionAnimCondition;
  }>;

export type RestartTrainingAction = Action<typeof constants.RESTART_TRAINING_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
    totalCount: number;
  }>;

export type StartExamAction = Action<typeof constants.START_EXAM_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
    totalCount: number;
  }>;

export type StartQuestionAction = Action<typeof constants.START_QUESTION_NAME> &
  Payload<{
    questionId: QuestionId;
    content: {
      yomiFuda: YomiFuda;
      toriFudaList: [ToriFuda, ToriFuda, ToriFuda, ToriFuda];
    };
  }>;

export type ResetQuestionAction = Action<typeof constants.RESET_QUESTION_NAME>;

export type FinishQuestionAction = Action<typeof constants.FINISH_QUESTION_NAME> &
  Payload<{
    correctCount: number;
    averageAnswerSecond: number;
    answerList: Array<{
      questionId: QuestionId;
      isCorrect: boolean;
      correctKaruta: Karuta;
    }>;
  }>;

export type AnswerQuestionAction = Action<typeof constants.ANSWER_QUESTION_NAME> &
  Payload<{
    isCorrect: boolean;
    selectedKarutaNo: KarutaNo;
    correctKaruta: Karuta;
  }>;

export type ConfirmCorrectAction = Action<typeof constants.CONFIRM_CORRECT_NAME>;

export type OpenNextQuestionAction = Action<typeof constants.OPEN_NEXT_QUESTION_NAME> &
  Payload<{
    currentQuestionId: QuestionId;
  }>;

export type Actions =
  | StartTrainingAction
  | RestartTrainingAction
  | StartExamAction
  | StartQuestionAction
  | ResetQuestionAction
  | FinishQuestionAction
  | AnswerQuestionAction
  | ConfirmCorrectAction
  | OpenNextQuestionAction;

export type ActionCreator = {
  /**
   * 指定した条件から問題を作成/保存してStartTrainingActionを返す
   *
   * @param rangeFrom 出題範囲From
   * @param rangeTo 出題範囲To
   * @param kimariji 決まり字条件
   * @param color 色条件
   * @param kamiNoKuStyle 上の句表示形式
   * @param shimoNoKuStyle 下の句表示形式
   * @param questionAnim 問題の表示速度
   *
   * @returns {StartTrainingAction}
   * @throws {Error} 指定した条件に当てはまる歌が無い場合
   */
  startTraining: (
    rangeFrom: RangeFromCondition,
    rangeTo: RangeToCondition,
    kimariji: KimarijiCondition,
    color: ColorCondition,
    kamiNoKuStyle: KarutaStyleCondition,
    shimoNoKuStyle: KarutaStyleCondition,
    questionAnim: QuestionAnimCondition
  ) => StartTrainingAction;

  /**
   * 回答済の問題の中から間違えた問題をもとに問題を作成/保存してRestartTrainingActionを返す
   *
   * @returns {RestartTrainingAction}
   * @throws {Error} 指定した条件に当てはまる歌が無い場合
   */
  restartTraining: () => RestartTrainingAction;

  /**
   * 全ての歌を対象に問題を作成/保存してStartExamActionを返す
   *
   * @returns {StartExamAction}
   * @throws {Error} 指定した条件に当てはまる歌が無い場合
   */
  startExam: () => StartExamAction;

  /**
   * 問題の回答を開始してStartQuestionActionを返す
   *
   * @returns {StartQuestionAction}
   * @throws {Error} 指定した問題がない場合
   */
  startQuestion: (
    currentQuestionId: QuestionId,
    kamiNoKuStyle: KarutaStyleCondition,
    shimoNoKuStyle: KarutaStyleCondition,
    startDate: Date
  ) => StartQuestionAction;

  /**
   * 問題の回答を判定してAnswerQuestionActionを返す
   *
   * @returns {AnswerQuestionAction}
   * @throws {Error} 指定した問題がない場合
   * @throws {Error} 問題の回答が開始されていない場合
   */
  answerQuestion: (currentQuestionId: QuestionId, toriFuda: ToriFuda, answerDate: Date) => AnswerQuestionAction;

  /**
   * ConfirmCorrectActionを返す
   *
   * @returns {ConfirmCorrectAction}
   */
  confirmCorrect: () => ConfirmCorrectAction;

  /**
   * 次の問題を取得してOpenNextQuestionActionを返す
   *
   * @returns {OpenNextQuestionAction}
   * @throws {Error} 次の問題がない場合
   */
  openNextQuestionAction: (currentQuestionId: QuestionId) => OpenNextQuestionAction;

  /**
   * 問題を初期状態に戻してResetQuestionActionを返す
   *
   * @returns {ResetQuestionAction}
   */
  resetQuestion: () => ResetQuestionAction;

  /**
   * 問題を終了して結果を見る
   *
   * @returns {FinishQuestionAction}
   */
  finishQuestion: () => FinishQuestionAction;
};

export type State = Readonly<{
  state: 'waiting' | 'ready' | 'playing' | 'confirm' | 'finished';
  totalCount: number;
  currentPosition: number;
  currentQuestion?: {
    questionId: QuestionId;
    content?: {
      yomiFuda: YomiFuda;
      toriFudaList: [ToriFuda, ToriFuda, ToriFuda, ToriFuda];
    };
    answer?: {
      isCorrect: boolean;
      selectedKarutaNo: KarutaNo;
      correctKaruta: Karuta;
    };
  };
  result?: {
    correctCount: number;
    averageAnswerSecond: number;
    answerList: Array<{
      questionId: QuestionId;
      isCorrect: boolean;
      correctKaruta: Karuta;
    }>;
  };
  trainingCondition: {
    rangeFrom: RangeFromCondition;
    rangeTo: RangeToCondition;
    kimariji: KimarijiCondition;
    color: ColorCondition;
    kamiNoKuStyle: KarutaStyleCondition;
    shimoNoKuStyle: KarutaStyleCondition;
    questionAnim: QuestionAnimCondition;
  };
}>;
