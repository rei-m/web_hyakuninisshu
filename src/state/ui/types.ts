import { Action } from 'redux';
import * as constants from './constants';
import { Color, Kimariji } from '@src/domain/models';
import { Payload } from '..';

export type KarutasFilter = Readonly<{
  colors: Array<{ color: Color; checked: boolean }>;
  kimarijis: Array<{ kimariji: Kimariji; checked: boolean }>;
}>;

export type OpenKarutasFilterAction = Action<typeof constants.OPEN_KARUTAS_FILTER_NAME>;

export type CloseKarutasFilterAction = Action<typeof constants.CLOSE_KARUTAS_FILTER_NAME>;

export type ToggleKarutasColorAction = Action<typeof constants.TOGGLE_KARUTAS_COLOR_NAME> &
  Payload<{
    color: Color;
    checked: boolean;
  }>;

export type ToggleKarutasKimarijiAction = Action<typeof constants.TOGGLE_KARUTAS_KIMARIJI_NAME> &
  Payload<{
    kimariji: Kimariji;
    checked: boolean;
  }>;

export type Actions =
  | OpenKarutasFilterAction
  | CloseKarutasFilterAction
  | ToggleKarutasColorAction
  | ToggleKarutasKimarijiAction;

export type ActionCreator = {
  /**
   * OpenKarutasFilterActionを返す
   *
   * @returns {OpenKarutasFilterAction}
   */
  openKarutasFilter: () => OpenKarutasFilterAction;

  /**
   * CloseKarutasFilterActionを返す
   *
   * @returns {CloseKarutasFilterAction}
   */
  closeKarutasFilter: () => CloseKarutasFilterAction;

  /**
   * ToggleKarutasColorActionを返す
   *
   * @param color 歌の色
   * @param checked 表示する/しない
   *
   * @returns {ToggleKarutasColorAction}
   */
  toggleKarutasColor: (color: Color, checked: boolean) => ToggleKarutasColorAction;

  /**
   * ToggleKarutasKimarijiActionを返す
   *
   * @param kimariji 歌の決まり字
   * @param checked 表示する/しない
   *
   * @returns {ToggleKarutasKimarijiAction}
   */
  toggleKarutasKimariji: (kimariji: Kimariji, checked: boolean) => ToggleKarutasKimarijiAction;
};

export type State = Readonly<{
  karutasFilter: KarutasFilter & { open: boolean };
}>;
