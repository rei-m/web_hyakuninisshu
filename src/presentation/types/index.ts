import { RouteComponentProps } from '@reach/router';

export type SiteMetaData = {
  title: string;
  description: string;
  author: string;
};

export type GeneratedPageComponentProps<T = {}> = { pageContext: T } & RouteComponentProps<T>;

export type AppErrorType = 'Network' | 'NotFound' | 'Unknown';

export class AppError implements Error {
  private _message: string;
  private _type: AppErrorType;

  constructor(message: string, type: AppErrorType) {
    this._message = message;
    this._type = type;
  }

  public get name() {
    return 'AppError';
  }

  public get message() {
    return this._message;
  }

  public get type() {
    return this._type;
  }

  public toString() {
    return `${this.name}: ${this.message}`;
  }
}
