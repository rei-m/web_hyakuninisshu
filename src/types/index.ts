export interface Karuta {
  id: number;
  imageNo: string;
}

export interface Question {
  correctKaruta: Karuta;
  choices: Karuta[];
}

export interface OptionItem<T> {
  readonly name: string;
  readonly value: T;
}
