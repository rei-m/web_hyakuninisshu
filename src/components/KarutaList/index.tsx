import * as React from 'react';
import styled from 'styled-components';
import { withAppTheme } from '../../styles';
import { Karuta } from '../../types';
import KarutaListRow from './KarutaListRow';

export interface KarutaListProps {
  readonly karutas: Karuta[];
  readonly onClickRow: (karutaId: number) => void;
}

const Root = withAppTheme(styled.div)`
  display: flex;
  flex-wrap: wrap;
  margin: auto;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  @media screen and (min-width: ${({ theme }) => theme.minWidthWide}) {
    flex-direction: row;
  }
`;

const KarutaList: React.SFC<KarutaListProps> = ({ karutas, onClickRow }) => (
  <Root>
    {karutas.map((k, i) => (
      <KarutaListRow
        karuta={k}
        key={i}
        onClickRow={onClickRow}
        data-test={`row-${k.id}`}
      />
    ))}
  </Root>
);

export default KarutaList;
