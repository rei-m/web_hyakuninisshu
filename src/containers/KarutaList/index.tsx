import * as React from 'react';
import { connect } from 'react-redux';
import styled from '@src/styles/styled-components';
import { GlobalState } from '@src/state';
import { uiSelectors } from '@src/state/ducks/ui';
import Row from '@src/containers/KarutaList/Row';
import { Karuta } from '@src/types';
import ErrorMessage from '@src/components/ErrorMessage';

export interface Props {
  karutas: Karuta[];
  onClickRow: (karutaNo: number) => void;
}

const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin: auto;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  @media screen and (min-width: ${({ theme }) => theme.minWidthWide}) {
    flex-direction: row;
  }
`;

export const KarutaList: React.FC<Props> = ({ karutas, onClickRow }) => (
  <Container>
    {karutas.length > 0 ? (
      karutas.map(karuta => (
        <Row karuta={karuta} key={karuta.id} onClickRow={onClickRow} data-test={`row-${karuta.no}`} />
      ))
    ) : (
      <ErrorMessage text="歌が見つかりませんでした。絞り込みを見直してください。" />
    )}
  </Container>
);

export const mapStateToProps = ({ ui }: GlobalState, { karutas, onClickRow }: Props): Props => ({
  karutas: uiSelectors.filterKarutas(karutas, ui.karutasFilter),
  onClickRow,
});

export default connect(mapStateToProps)(KarutaList);
