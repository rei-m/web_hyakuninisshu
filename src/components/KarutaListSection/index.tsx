import * as React from 'react';
import styled from 'styled-components';
import { withAppTheme } from '../../styles';
import PageTitle from '../PageTitle';
import KarutaList from '../KarutaList';
import { Karuta } from '../../types';

export interface KarutaListSectionProps {
  readonly karutas: Karuta[];
  readonly onClickKarutaListRow: (karutaId: number) => void;
}

const RootSection = withAppTheme(styled.section)`
  box-sizing: border-box;
  padding: ${({ theme }) => theme.spacing2x};
`;

const KarutaListSection: React.SFC<KarutaListSectionProps> = ({
  karutas,
  onClickKarutaListRow
}) => (
  <RootSection>
    <PageTitle title="歌一覧" />
    <KarutaList karutas={karutas} onClickRow={onClickKarutaListRow} />
  </RootSection>
);

export default KarutaListSection;
