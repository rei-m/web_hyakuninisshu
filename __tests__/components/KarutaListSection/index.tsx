import * as ReactTestRenderer from 'react-test-renderer';
import * as React from 'react';
import { shallow } from 'enzyme';
import { create } from '../../factories';
import KarutaListSection, {
  KarutaListSectionProps
} from '@src/components/KarutaListSection';
import KarutaList from '@src/components/KarutaList';
import { Karuta } from '@src/types';

describe('<KarutaListSection />', () => {
  let props: KarutaListSectionProps;

  beforeEach(() => {
    props = {
      karutas: [
        create<Karuta>('karuta', {
          id: 1
        }),
        create<Karuta>('karuta', {
          id: 2
        })
      ],
      onClickKarutaListRow: jest.fn()
    };
  });

  it('should render component', () => {
    const renderer = ReactTestRenderer.create(<KarutaListSection {...props} />);
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should fire onClickKarutaListRow when KarutaList clicked', () => {
    const wrapper = shallow(<KarutaListSection {...props} />);
    wrapper
      .find(KarutaList)
      .props()
      .onClickRow(1);
    expect(props.onClickKarutaListRow).toHaveBeenCalledWith(1);
  });
});
