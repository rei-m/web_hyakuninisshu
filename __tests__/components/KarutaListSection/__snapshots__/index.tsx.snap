// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KarutaListSection /> should render component 1`] = `
<section
  className="sc-htoDjs gFSeai"
>
  <h1
    className="sc-bdVaJa fXsTJn"
  >
    歌一覧
  </h1>
  <div
    className="sc-gzVnrw iKENGW"
  >
    <div
      className="sc-bwzfXH epKFfr"
      onClick={[Function]}
    >
      <div
        className="sc-htpNat btqCfC"
      >
        <div
          className="sc-ifAKCX gXBqWv"
        >
          一番
        </div>
        <img
          alt="一番"
          src="karuta_001.jpg"
          style={
            Object {
              "width": 40,
            }
          }
        />
      </div>
      <div
        className="sc-bxivhb jdZxpC"
      >
        <div
          className="sc-EHOje kLbaUi"
        >
          初句 二句 三句
        </div>
        <div
          className="sc-EHOje kLbaUi"
        >
          四句 五句
        </div>
        <div
          className="sc-bZQynM fvXZAg"
        >
          詠み手
        </div>
      </div>
    </div>
    <div
      className="sc-bwzfXH epKFfr"
      onClick={[Function]}
    >
      <div
        className="sc-htpNat btqCfC"
      >
        <div
          className="sc-ifAKCX gXBqWv"
        >
          二番
        </div>
        <img
          alt="二番"
          src="karuta_002.jpg"
          style={
            Object {
              "width": 40,
            }
          }
        />
      </div>
      <div
        className="sc-bxivhb jdZxpC"
      >
        <div
          className="sc-EHOje kLbaUi"
        >
          初句 二句 三句
        </div>
        <div
          className="sc-EHOje kLbaUi"
        >
          四句 五句
        </div>
        <div
          className="sc-bZQynM fvXZAg"
        >
          詠み手
        </div>
      </div>
    </div>
  </div>
</section>
`;
